subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java

```diff
@@ -538,4 +540,27 @@ public void setAnnotationProcessorGeneratedSourcesDirectory(Provider<File> file)
         this.annotationProcessorGeneratedSourcesDirectory.set(file);
     }
 
+    /**
+     * Returns the directory where JNI-generated headers will be put.
+     *
+     * @since 4.10
+     */
+    @Incubating
+    @Nullable
+    @Optional
+    @OutputDirectory
+    public File getHeaderOutputDirectory() {
+        return headerOutputDirectory;
+    }
+
+    /**
+     * Sets the directory where JNI-generated headers will be put.
+     *
+     * @since 4.10
+     */
+    @Incubating
+    public void setHeaderOutputDirectory(File headerOutputDirectory) {
```

Comment:
Should the parameter have `@Nullable`, too?

Reply:
Should we use a `Provider<File>`?

Reply:
Yes, it should be nullable. 

If we use the Property API, then there should only be a single getter of type `DirectoryProperty`. What we have for `annotationProcessorPath` above is an anti-pattern. I opted for the simple `File` because all other fields in this class are not using the `Property` API, so it felt a bit odd. But we could certainly start here. @big-guy what's our current plan, should we use `Property`s for any new properties are should we do a more consistent restructuring later?